### PR TITLE
New resource: aws_autoscalingplans_scaling_plan

### LINKED ISSUE #6562: Feature Request: Auto Scaling Plans (Dynamic/Predictive Auto Scaling Groups)
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Amazon recently introduced a predictive scaling capability in auto scaling groups. Auto Scaling Plans use dynamic scaling and predictive scaling to automatically scale ASGs. I believe a new resource is required to support this.

### New Resource(s)

* `aws_autoscaling_plan`

### Potential Terraform Configuration

```hcl
resource "aws_autoscaling_plan" "this" {
  name                    = "this-autoscaling-plan"
  min_capacity            = 1
  max_capacity            = 7
  predictive_scaling_mode = "forecast-and-scale"
  scalable_dimension      = "autoscaling:autoScalingGroup:DesiredCapacity"
  autoscaling_group_name  = "${aws_autoscaling_group.that.name}"
}
```

### References

* [AWS News Blog: New – Predictive Scaling for EC2, Powered by Machine Learning](https://aws.amazon.com/blogs/aws/new-predictive-scaling-for-ec2-powered-by-machine-learning/)
* [API Reference: CreateScalingPlan](https://docs.aws.amazon.com/autoscaling/plans/APIReference/API_CreateScalingPlan.html)
* [What Is AWS Auto Scaling? (Updated)](https://docs.aws.amazon.com/autoscaling/plans/userguide/what-is-aws-auto-scaling.html)

### PR DESCRIPTION
**Pull request by ewbankkit**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Fixes #6562

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
FEATURES:

* New Resource: aws_autoscalingplans_scaling_plan (#6562)
```

Output from acceptance testing:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsAutoScalingPlansScalingPlan_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAwsAutoScalingPlansScalingPlan_ -timeout 120m
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_AlbAsgDynamoDbEcsDynamicScaling
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_AlbAsgDynamoDbEcsDynamicScaling
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_CustomMetrics
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_CustomMetrics
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_AlbAsgDynamoDbEcsDynamicScaling
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_CustomMetrics
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_CustomMetrics (75.44s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling (80.35s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling (80.66s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicUpdate (108.83s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_AlbAsgDynamoDbEcsDynamicScaling (296.77s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	296.838s
```

### CODE REVIEW

**Comment by ewbankkit** [comment]
Currently includes the commit from https://github.com/terraform-providers/terraform-provider-aws/pull/8962.

**Comment by ewbankkit** [comment]
The name **aws_autoscalingplans_scaling_plan** is just a straw man as the resource is implemented via a new `autoscalingplans` AWS API and CloudFormation names its [equivalent resource](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-autoscalingplans-scalingplan.html) similarly.
All the AWS documentation and AWS console just calls them **Auto Scaling scaling plans* so `aws_autoscaling_scaling_plan` may be the better name and also consolidate the documentation into the _Autoscaling_ section rather than a separate section.

**Comment by ewbankkit** [comment]
Removing _WIP_.
I've actually come to be OK with the name `aws_autoscalingplans_scaling_plan`.

**Comment by ewbankkit** [comment]
Rebased to fix merge conflict.

**Comment by ewbankkit** [comment]
Rebased to fix merge conflict.
Acceptance tests:
```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsAutoScalingPlansScalingPlan_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccAwsAutoScalingPlansScalingPlan_ -timeout 120m
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling (80.60s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling (81.03s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicUpdate (112.85s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	112.920s
``

**Comment by ewbankkit** [comment]
Rebased and did the Terraform Plugin SDK migration for the new resource.
Re-ran acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsAutoScalingPlansScalingPlan_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccAwsAutoScalingPlansScalingPlan_ -timeout 120m
go: finding github.com/terraform-providers/terraform-provider-tls v2.1.1+incompatible
go: finding github.com/terraform-providers/terraform-provider-tls v2.1.1+incompatible
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling (76.66s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling (77.07s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicUpdate (104.08s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	104.122s
```

File: aws/validators.go [hunk 1]
```diff
@@ -63,6 +63,26 @@ func validateStringNotMatch(r *regexp.Regexp, message string) schema.SchemaValid
 	}
 }
 
+// validateStringDoesNotContainAny returns a SchemaValidateFunc which validates that the
```
Comment (ewbankkit) [inline aws/validators.go]: Track https://github.com/hashicorp/terraform-plugin-sdk/pull/212.

**Comment by ewbankkit** [comment]
### TODO

* Any documentation changes for https://github.com/terraform-providers/terraform-provider-aws/pull/10684
* Track https://github.com/terraform-providers/terraform-provider-aws/pull/10768 to use `StringDoesNotContainAny`

**Comment by jcruzfrontline** [comment]
@bflad - This would be great to get this merged. We would like to utilize this feature in our staging & production environment without manual steps. Thank you for your hard work on this @ewbankkit !

**Comment by mr-ransel** [comment]
I heartily second @jcruzfrontline . We are getting ready to fork and rebuild terraform internally to make use of this change ourselves, just because of the inherent danger in having to manage predictive scaling plans by hand.

**Comment by darmawanaris** [comment]
Hi @ewbankkit , why this PR is not merged yet ?

**Comment by ewbankkit** [comment]
Revisit, considering https://github.com/terraform-providers/terraform-provider-aws/pull/13194.

File: aws/resource_aws_autoscalingplans_scaling_plan.go [hunk 2]
```diff
@@ -0,0 +1,1081 @@
...
+	}
+	if v, ok := mTargetTrackingConfiguration["target_value"].(float64); ok {
+		buf.WriteString(fmt.Sprintf("%g-", v))
+	}
+
+	return hashcode.String(buf.String())
+}
+
+// stableMapHash returns a stable hash value for a map.
+func stableMapHash(m map[string]interface{}) int {
```
Comment (ewbankkit) [inline aws/resource_aws_autoscalingplans_scaling_plan.go:1067]: Verify whether this is needed after merge of https://github.com/terraform-providers/terraform-provider-aws/pull/12740, pulling in https://github.com/hashicorp/terraform-plugin-sdk/pull/362.

Comment (ewbankkit) [reply aws/resource_aws_autoscalingplans_scaling_plan.go]: Removed custom set hash functions.

**Comment by ewbankkit** [comment]
Re-ran acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsAutoScalingPlansScalingPlan_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccAwsAutoScalingPlansScalingPlan_ -timeout 120m
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_disappears
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_disappears
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_disappears
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_disappears (53.23s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling (57.27s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicUpdate (80.78s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling (81.58s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	81.641s
```

**Comment by ewbankkit** [comment]
Rebase to fix merge conflicts and simplify acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsAutoScalingPlansScalingPlan_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccAwsAutoScalingPlansScalingPlan_ -timeout 120m
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_disappears
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_disappears
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_disappears
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_disappears (52.61s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicUpdate (79.53s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling (81.24s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling (86.21s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	86.280s
```

**Comment by maitdaddy** [comment]
@ewbankkit We salute you for all your efforts on this request!

**Comment by jcruzfrontline** [comment]
@ewbankkit  - I second that!

**Comment by ewbankkit** [comment]
This is on [this quarter's road map](https://github.com/terraform-providers/terraform-provider-aws/blob/3713ea9f8ce8f3bb1fd572beed8b1fa1967af7ff/ROADMAP.md#issues-and-enhancements).

**Comment by kyprifog** [comment]
Excited to see this one make it in

**Comment by ewbankkit** [comment]
Rebased and upgraded new code to Plugin SDK v2.
Re-ran acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsAutoScalingPlansScalingPlan_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsAutoScalingPlansScalingPlan_ -timeout 120m
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_disappears
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_disappears
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_disappears
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_disappears
    resource_aws_autoscalingplans_scaling_plan_test.go:268: [INFO] Got non-empty plan, as expected
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling (52.75s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling (61.57s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_disappears (61.58s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicUpdate (83.10s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	83.157s
```

**Comment by ssabo** [comment]
I am not sure who to direct this question to. What is preventing this patch from being merged currently?

**Review by anGie44 — CHANGES_REQUESTED** [review]
Thank you for this PR @ewbankkit! Design looks great overall. It seems though the waiter methods may need some tweaking in the `Pending` states values as the following tests are consistently failing in Commercial and Gov Cloud:

```
--- FAIL: TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling (9.54s)
--- FAIL: TestAccAwsAutoScalingPlansScalingPlan_disappears (9.62s)
--- FAIL: TestAccAwsAutoScalingPlansScalingPlan_basicUpdate (9.66s)
--- FAIL: TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling (36.08s)
```

with the error(s):
```
Error: error waiting for Auto Scaling Scaling Plan (tf-acc-test-6121015609472249022/1) to be created: unexpected state 'NotFound', wanted target 'Active, ActiveWithProblems'. last error: %!s(<nil>)
```
```
Error: error waiting for Auto Scaling Scaling Plan (tf-acc-test-6406453948128485035/1) to be deleted: unexpected state 'Active', wanted target 'NotFound'. last error: %!s(<nil>)
```

**Comment by ewbankkit** [comment]
Modified the status codes used by the create and delete waiters to better handle eventual consistency.

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsAutoScalingPlansScalingPlan_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsAutoScalingPlansScalingPlan_ -timeout 120m
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_disappears
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_disappears
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_disappears
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_disappears
    resource_aws_autoscalingplans_scaling_plan_test.go:257: [INFO] Got non-empty plan, as expected
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_disappears (57.76s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling (60.68s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling (61.08s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicUpdate (93.65s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	93.687s
```

File: website/docs/r/autoscalingplans_scaling_plan.html.markdown [hunk 3]
```diff
@@ -11,6 +11,10 @@ description: |-
 Manages an AWS Auto Scaling scaling plan.
 More information can be found in the [AWS Auto Scaling User Guide](https://docs.aws.amazon.com/autoscaling/plans/userguide/what-is-aws-auto-scaling.html).
 
+~> **NOTE:** The AWS Auto Scaling service uses an AWS IAM service-linked role to manage predictive scaling of Amazon EC2 Auto Scaling groups. The service attempts to automatically create this role the first time a scaling plan with predictive scaling enabled is created.
+An [`aws_iam_service_linked_role`](/docs/providers/aws/r/iam_service_linked_role.html) resource can be used to manually managed this role.
```
Comment (anGie44) [inline website/docs/r/autoscalingplans_scaling_plan.html.markdown:15]: small correction:  
```suggestion
An [`aws_iam_service_linked_role`](/docs/providers/aws/r/iam_service_linked_role.html) resource can be used to manually manage this role.
```

**Comment by ewbankkit** [comment]
Added `testAccPreCheckIamServiceLinkedRole` and used if for those acceptance tests that use predictive scaling.
Without the service-linked role:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsAutoScalingPlansScalingPlan_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsAutoScalingPlansScalingPlan_ -timeout 120m
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_disappears
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_disappears
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_disappears
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
    provider_test.go:601: skipping tests; missing IAM service-linked role /aws-service-role/autoscaling-plans. Please create the role and retry
--- SKIP: TestAccAwsAutoScalingPlansScalingPlan_basicUpdate (0.87s)
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
    provider_test.go:601: skipping tests; missing IAM service-linked role /aws-service-role/autoscaling-plans. Please create the role and retry
--- SKIP: TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling (0.98s)
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_disappears
    resource_aws_autoscalingplans_scaling_plan_test.go:263: [INFO] Got non-empty plan, as expected
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_disappears (35.14s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling (59.48s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	59.523s
```

after [manually creating the service-linked role](https://docs.aws.amazon.com/autoscaling/plans/userguide/aws-auto-scaling-service-linked-roles.html#create-service-linked-role-manual):

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsAutoScalingPlansScalingPlan_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsAutoScalingPlansScalingPlan_ -timeout 120m
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== RUN   TestAccAwsAutoScalingPlansScalingPlan_disappears
=== PAUSE TestAccAwsAutoScalingPlansScalingPlan_disappears
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_disappears
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicUpdate
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling
=== CONT  TestAccAwsAutoScalingPlansScalingPlan_disappears
    resource_aws_autoscalingplans_scaling_plan_test.go:263: [INFO] Got non-empty plan, as expected
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_disappears (42.35s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling (56.60s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling (64.38s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicUpdate (78.71s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	78.746s
```

File: aws/internal/service/autoscalingplans/finder/finder.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package finder
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/autoscalingplans"
+)
+
+// ScalingPlan returns the scaling plan corresponding to the specified name and version.
+// Returns returns nil and potentially an API error if no scaling plan is found.
```
Comment (anGie44) [inline aws/internal/service/autoscalingplans/finder/finder.go:9]: Nit: 
```suggestion
// Returns nil and potentially an API error if no scaling plan is found.
```

**Review by anGie44 — APPROVED** [review]
Looks great and acceptance tests have been corrected 👍 

Output of acceptance tests (Commercial):
```
--- PASS: TestProvider_impl (0.12s)
--- PASS: TestProvider (1.81s)
--- PASS: TestAccAWSProvider_Region_AwsGovCloudUs (36.00s)
--- PASS: TestAccAWSProvider_Region_AwsCommercial (36.88s)
--- PASS: TestAccAWSProvider_Region_AwsChina (37.02s)
--- PASS: TestAccAWSProvider_IgnoreTags_KeyPrefixes_None (46.29s)
--- PASS: TestAccAWSProvider_Endpoints (46.14s)
--- PASS: TestAccAWSProvider_IgnoreTags_Keys_One (46.60s)
--- PASS: TestAccAWSProvider_IgnoreTags_EmptyConfigurationBlock (46.34s)
--- PASS: TestAccAWSProvider_IgnoreTags_KeyPrefixes_One (46.83s)
--- PASS: TestAccAWSProvider_IgnoreTags_Keys_None (46.92s)
--- PASS: TestAccAWSProvider_IgnoreTags_KeyPrefixes_Multiple (47.03s)
--- PASS: TestAccAWSProvider_IgnoreTags_Keys_Multiple (47.26s)
--- PASS: TestAccAWSProvider_AssumeRole_Empty (48.30s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_disappears (73.48s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling (92.06s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling (92.39s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicUpdate (117.60s)
```
Output of acceptance tests (Gov Cloud):
```
--- PASS: TestProvider_impl (0.17s)
--- PASS: TestProvider (1.17s)
--- SKIP: TestAccAwsAutoScalingPlansScalingPlan_basicPredictiveScaling (0.88s)
--- SKIP: TestAccAwsAutoScalingPlansScalingPlan_basicUpdate (0.56s)
--- PASS: TestAccAWSProvider_Region_AwsChina (34.51s)
--- PASS: TestAccAWSProvider_Region_AwsGovCloudUs (34.30s)
--- PASS: TestAccAWSProvider_Region_AwsCommercial (35.68s)
--- PASS: TestAccAWSProvider_Endpoints (43.40s)
--- PASS: TestAccAWSProvider_IgnoreTags_Keys_Multiple (42.95s)
--- PASS: TestAccAWSProvider_IgnoreTags_Keys_None (43.51s)
--- PASS: TestAccAWSProvider_IgnoreTags_Keys_One (43.36s)
--- PASS: TestAccAWSProvider_IgnoreTags_KeyPrefixes_One (44.37s)
--- PASS: TestAccAWSProvider_AssumeRole_Empty (44.08s)
--- PASS: TestAccAWSProvider_IgnoreTags_KeyPrefixes_None (44.39s)
--- PASS: TestAccAWSProvider_IgnoreTags_KeyPrefixes_Multiple (44.28s)
--- PASS: TestAccAWSProvider_IgnoreTags_EmptyConfigurationBlock (44.40s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_basicDynamicScaling (67.52s)
--- PASS: TestAccAwsAutoScalingPlansScalingPlan_disappears (71.39s)
```

**Pull request merged**